File: engine/table/src/main/java/io/deephaven/stream/TablePublisher.java
package io.deephaven.stream;

import io.deephaven.engine.context.ExecutionContext;
import io.deephaven.engine.table.Table;
import io.deephaven.engine.table.TableDefinition;
import io.deephaven.engine.table.impl.sources.ArrayBackedColumnSource;
import io.deephaven.engine.updategraph.UpdateGraph;
import io.deephaven.engine.util.input.InputTableStatusListener;
import io.deephaven.engine.util.input.InputTableUpdater;
import io.deephaven.util.annotations.TestUseOnly;

import javax.annotation.Nullable;
import java.util.Map;
import java.util.Objects;
import java.util.function.Consumer;

/**
 * Produces a {@link Table#BLINK_TABLE_ATTRIBUTE blink table} from {@link #add(Table) added tables}.
 */
public class TablePublisher {

    /**
     * Constructs a table publisher.
     *
     * <p>
     * The {@code onFlushCallback}, if present, is called once at the beginning of each update graph cycle. This is a
     * pattern that allows publishers to add any data they may have been batching. Do note though, this blocks the
     * update cycle from proceeding, so implementations should take care to not do extraneous work.
     *
     * <p>
     * The {@code onShutdownCallback}, if present, is called one time when the publisher should stop publishing new data
     * and release any related resources as soon as practicable since publishing won't have any downstream effects.
     *
     * <p>
     * Equivalent to calling {@link #of(String, TableDefinition, Consumer, Runnable, UpdateGraph, int)} with the
     * {@code updateGraph} from {@link ExecutionContext#getContext()} and {@code chunkSize}
     * {@value ArrayBackedColumnSource#BLOCK_SIZE}.
     *
     * @param name the name
     * @param definition the table definition
     * @param onFlushCallback the on-flush callback
     * @param onShutdownCallback the on-shutdown callback
     * @return the table publisher
     */
    public static TablePublisher of(
            String name,
            TableDefinition definition,
            @Nullable Consumer<TablePublisher> onFlushCallback,
            @Nullable Runnable onShutdownCallback) {
        return of(name, definition, onFlushCallback, onShutdownCallback, ExecutionContext.getContext().getUpdateGraph(),
                ArrayBackedColumnSource.BLOCK_SIZE);
    }

    /**
     * Constructs a table publisher.
     *
     * <p>
     * The {@code onFlushCallback}, if present, is called once at the beginning of each update graph cycle. This is a
     * pattern that allows publishers to add any data they may have been batching. Do note though, this blocks the
     * update cycle from proceeding, so implementations should take care to not do extraneous work.
     *
     * <p>
     * The {@code onShutdownCallback}, if present, is called one time when the publisher should stop publishing new data
     * and release any related resources as soon as practicable since publishing won't have any downstream effects.
     *
     * <p>
     * The {@code chunkSize} is the size at which chunks will be filled from the source table during an
     * {@link TablePublisher#add}. The suggested value is {@value ArrayBackedColumnSource#BLOCK_SIZE}.
     *
     * @param name the name
     * @param definition the table definition
     * @param onFlushCallback the on-flush callback
     * @param onShutdownCallback the on-shutdown callback
     * @param updateGraph the update graph for the blink table
     * @param chunkSize the chunk size is the maximum size
     * @return the table publisher
     */
    public static TablePublisher of(
            String name,
            TableDefinition definition,
            @Nullable Consumer<TablePublisher> onFlushCallback,
            @Nullable Runnable onShutdownCallback,
            UpdateGraph updateGraph,
            int chunkSize) {
        final TablePublisher[] publisher = new TablePublisher[1];
        final TableStreamPublisherImpl impl =
                new TableStreamPublisherImpl(name, definition,
                        onFlushCallback == null ? null : () -> onFlushCallback.accept(publisher[0]), onShutdownCallback,
                        chunkSize);
        final StreamToBlinkTableAdapter adapter =
                new StreamToBlinkTableAdapter(definition, impl, updateGraph, name, Map.of(), false);
        publisher[0] = new TablePublisher(impl, adapter);
        adapter.initialize();
        return publisher[0];
    }

    private final TableStreamPublisherImpl publisher;
    private final StreamToBlinkTableAdapter adapter;

    private TablePublisher(TableStreamPublisherImpl publisher, StreamToBlinkTableAdapter adapter) {
        this.publisher = Objects.requireNonNull(publisher);
        this.adapter = Objects.requireNonNull(adapter);
    }

    /**
     * The {@link #table blink table's} definition.
     *
     * @return the definition
     */
    public TableDefinition definition() {
        return publisher.definition();
    }

    /**
     * The {@link Table#BLINK_TABLE_ATTRIBUTE blink table}.
     *
     * <p>
     * May return {@code null} if invoked more than once and the initial caller does not enforce strong reachability of
     * the result.
     *
     * @return the blink table
     */
    public Table table() {
        return adapter.table();
    }

    /**
     * Adds a snapshot of the data from {@code table} into the {@link #table blink table}. The added {@code table} must
     * contain a superset of the columns from the {@link #definition() definition}; the columns may be in any order.
     * Columns from {@code table} that are not in the {@link #definition() definition} are ignored.
     *
     * <p>
     * All of the data from {@code table} will be:
     *
     * <ol>
     * <li><b>consistent</b> with a point in time</li>
     * <li><b>fully contained</b> in a single blink table's update cycle</li>
     * <li><b>non-interleaved</b> with any other calls to add (concurrent, or not)</li>
     * </ol>
     *
     * @param table the table to add
     */
    public void add(Table table) {
        publisher.add(table);
    }

    /**
     * Indicate that data publication has failed. {@link #table() Blink table}
     * {@link io.deephaven.engine.table.TableListener listeners} will be notified of the failure, the on-shutdown
     * callback will be invoked if it hasn't already been, {@code this} publisher will no longer be {@link #isAlive()
     * alive}, and future calls to {@link #add(Table) add} will silently return without publishing. These effects may
     * resolve asynchronously.
     *
     * @param failure the failure
     */
    public void publishFailure(Throwable failure) {
        publisher.publishFailure(failure);
    }

    /**
     * Checks whether {@code this} is alive; if {@code false}, the caller should stop adding new data and release any
     * related resources as soon as practicable since adding data won't have any downstream effects.
     *
     * <p>
     * Once this is {@code false}, it will always remain {@code false}. For more prompt notifications, callers may
     * prefer to use on-shutdown callbacks.
     *
     * @return if this is alive
     */
    public boolean isAlive() {
        return adapter.isAlive();
    }

    /**
     * Creates a new {@link Table#BLINK_TABLE_ATTRIBUTE blink table} with its {@link Table#getAttribute(String)
     * attribute} {@value Table#INPUT_TABLE_ATTRIBUTE} set to an {@link InputTableUpdater} implementation based on
     * {@code this}. The implementation's definition of "completed" with respect to {@link InputTableUpdater#add(Table)}
     * and {@link InputTableUpdater#addAsync(Table, InputTableStatusListener)} matches the semantics provided by
     * {@link #add(Table)} - that is, "completed" means that a snapshot of {@code newData} has been taken and handed
     * off. The implementation does not implement {@link InputTableUpdater#delete(Table)} nor
     * {@link InputTableUpdater#deleteAsync(Table, InputTableStatusListener)}.
     *
     * <p>
     * May return {@code null} if invoked more than once and the initial caller does not enforce strong reachability of
     * the result.
     *
     * @return the input-table blink table
     */
    public Table inputTable() {
        final Table table = adapter.table();
        if (table == null) {
            return null;
        }
        return table.withAttributes(Map.of(Table.INPUT_TABLE_ATTRIBUTE, publisher.inputTableUpdater()));
    }

    @TestUseOnly
    void runForUnitTests() {
        adapter.run();
    }
}

File: py/server/deephaven/stream/table_publisher.py
#
# Copyright (c) 2016-2023 Deephaven Data Labs and Patent Pending
#
"""The table_publisher module supports publishing Deephaven Tables into blink Tables."""

import jpy

from typing import Callable, Dict, Optional, Tuple

from deephaven._wrapper import JObjectWrapper
from deephaven.column import Column
from deephaven.dtypes import DType
from deephaven.execution_context import get_exec_ctx
from deephaven.jcompat import j_lambda, j_runnable
from deephaven.table import Table
from deephaven.update_graph import UpdateGraph

_JTableDefinition = jpy.get_type("io.deephaven.engine.table.TableDefinition")
_JTablePublisher = jpy.get_type("io.deephaven.stream.TablePublisher")
_JRuntimeException = jpy.get_type("java.lang.RuntimeException")
_JConsumer = jpy.get_type("java.util.function.Consumer")


class TablePublisher(JObjectWrapper):
    """The interface for publishing table data into a blink table."""

    j_object_type = _JTablePublisher

    def __init__(self, j_table_publisher: jpy.JType):
        self.j_table_publisher = j_table_publisher

    @property
    def j_object(self) -> jpy.JType:
        return self.j_table_publisher

    @property
    def is_alive(self) -> bool:
        """Checks whether this is alive; if False, the caller should stop adding new data and release any related
        resources as soon as practicable since adding new data won't have any downstream effects.

        Once this is False, it will always remain False. For more prompt notifications, callers may prefer to use
        on_shutdown_callback during construction.

        Returns:
            if this publisher is alive
        """
        return self.j_table_publisher.isAlive()

    def add(self, table: Table) -> None:
        """Adds a snapshot of the data from table into the blink table. The table must contain a superset of the columns
        from the blink table's definition; the columns may be in any order. Columns from table that are not in the blink
        table's definition are ignored.

        All of the data from table will be:

        1. consistent with a point in time
        2. fully contained in a single blink table's update cycle
        3. non-interleaved with any other calls to add (concurrent, or not)

        Args:
            table (Table): the table to add
        """
        self.j_table_publisher.add(table.j_table)

    def publish_failure(self, failure: Exception) -> None:
        """Indicate that data publication has failed. Blink table listeners will be notified of the failure, the
        on-shutdown callback will be invoked if it hasn't already been, this publisher will no longer be alive, and
        future calls to add will silently return without publishing. These effects may resolve asynchronously.

        Args:
            failure (Exception): the failure
        """
        self.j_table_publisher.publishFailure(_JRuntimeException(str(failure)))


def table_publisher(
    name: str,
    col_defs: Dict[str, DType],
    on_flush_callback: Optional[Callable[[TablePublisher], None]] = None,
    on_shutdown_callback: Optional[Callable[[], None]] = None,
    update_graph: Optional[UpdateGraph] = None,
    chunk_size: int = 2048,
) -> Tuple[Table, TablePublisher]:
    """Constructs a blink Table and TablePublisher to populate it.

    Args:
        name (str): the name, used for logging
        col_defs (Dict[str, DType]): the column definitions for the resulting blink table
        on_flush_callback (Optional[Callable[[TablePublisher], None]]): the on-flush callback, if present, is called
            once at the beginning of each update graph cycle. This is a pattern that allows publishers to add any data
            they may have been batching. Do note though, this blocks the update cycle from proceeding, so
            implementations should take care to not do extraneous work.
        on_shutdown_callback (Optional[Callable[[], None]]): the on-shutdown callback, if present, is called one time
            when the caller should stop adding new data and release any related resources as soon as practicable since
            adding data won't have any downstream effects
        update_graph (Optional[UpdateGraph]): the update graph the resulting table will belong to. If unset, the update
            graph of the current execution context will be used.
        chunk_size (int): the chunk size, the size at which chunks will be filled from the source table during an add,
            defaults to 2048

    Returns:
        a two-tuple, where the first item is resulting blink Table and the second item is the TablePublisher
    """

    def adapt_callback(_table_publisher: jpy.JType):
        on_flush_callback(TablePublisher(j_table_publisher=_table_publisher))

    j_table_publisher = _JTablePublisher.of(
        name,
        _JTableDefinition.of(
            [
                Column(name=name, data_type=dtype).j_column_definition
                for name, dtype in col_defs.items()
            ]
        ),
        j_lambda(adapt_callback, _JConsumer, None) if on_flush_callback else None,
        j_runnable(on_shutdown_callback) if on_shutdown_callback else None,
        (update_graph or get_exec_ctx().update_graph).j_update_graph,
        chunk_size,
    )
    # Note: this differs from the java version where the end user is given access to TablePublisher.table().
    # In a theoretical world where java allows more ergonomic creation of tuples / tuple-typing, it would look more like
    # this python version.
    return (
        Table(j_table=j_table_publisher.table()),
        TablePublisher(j_table_publisher=j_table_publisher),
    )
